src/lib.rs

```diff
@@ -1,384 +1,73 @@
 #![cfg_attr(feature = "nightly",
-    feature(external_doc),
+    feature(doc_cfg, external_doc),
     doc(include = "../README.md"),
 )]
+#![cfg_attr(feature = "specialization",
+    feature(specialization),
+)]
 
 #[macro_use]
 extern crate require_unsafe_in_body;
 
 #[macro_use]
 mod utils;
-#[path = "utils/prelude.rs"]
-mod prelude;
+pub mod prelude {
+    #[doc(no_inline)]
+    pub use crate::{
+        extension_traits::{
+            AsOut,
+            ManuallyDropMut,
+            VecReserveUninit,
+        },
+        out_references::{
+            Out,
+            OutSlice,
+        },
+        uninit_array,
+    };
+}
 
 use_prelude!();
 
-mod impls;
+pub
+mod extension_traits;
+
+pub
+mod read;
 
-pub use self::traits::{
-    MaybeUninitExt,
-    slice::{
-        InitWithCopyFromSlice,
-    },
-    vec::{
-        VecExtendFromReader,
-        VecReserveUninit,
-    },
-};
-mod traits;
+pub
+mod out_references;
 
 #[doc(hidden)]
 pub use ::core;
 
 /// Sets up an inline / stack-allocated array of
-/// [uninitialized bytes][`MaybeUninit`].
+/// [uninitialized][`MaybeUninit`] elements.
 ///
 /// # Example
 ///
 /// ```rust
-/// use ::uninit::{ReadIntoUninit, uninit_byte_array};
+/// use ::uninit::{prelude::*, read::ReadIntoUninit};
 ///
 /// let mut reader = &b"Hello, World!"[..];
-/// let mut array = uninit_byte_array![_; 4]; // : [MaybeUninit<u8>; 4]
+/// let mut backing_array = uninit_array![u8; 4]; // : [MaybeUninit<u8>; 4]
+/// let buf = OutSlice::from(&mut backing_array[..]);
 /// assert_eq!(
-///     reader.read_into_uninit_exact(&mut array).unwrap(),
+///     reader.read_into_uninit_exact(buf).unwrap(),
 ///     b"Hell",
 /// );
 /// ```
 #[macro_export]
-macro_rules! uninit_byte_array {(
-    _; $count:expr
+macro_rules! uninit_array {(
+    $T:ty ; $count:expr
 ) => ({
     use $crate::core::mem::MaybeUninit;
-    type Array = [MaybeUninit<u8>; $count];
+    const __UNINIT_ARRAY_COUNT__: usize = $count;
```

Comment:
Let's allow `_` type inference at the call site, but keep the `$count` evaluation outside `unsafe {}` hygiene